numba/np/npy_randomgen.py

```diff
@@ -0,0 +1,1130 @@
+        idx = r & 0xff
+        sign = (r >> 8) & 0x1
+        rabs = (r >> 9) & 0x0007fffff
+        x = rabs * wi_float[idx]
+        if (sign & 0x1):
+            x = -x
+        if (rabs < ki_float[idx]):
+            return x
+        if (idx == 0):
+            while 1:
+                xx = -ziggurat_nor_inv_r_f * np.log1p(-next_float(bitgen))
+                yy = -np.log1p(-next_float(bitgen))
+                if (yy + yy > xx * xx):
+                    if ((rabs >> 8) & 0x1):
+                        return -(ziggurat_nor_r_f + xx)
+                    else:
+                        return ziggurat_nor_r_f + xx
+        else:
+            if (((fi_float[idx - 1] - fi_float[idx]) * next_float(bitgen) +
+                 fi_float[idx]) < np.exp(-0.5 * x * x)):
+                return x
+
+
+@register_jitable
+def random_standard_exponential(bitgen):
+    while 1:
+        ri = next_uint64(bitgen)
+        ri >>= 3
+        idx = ri & 0xFF
+        ri >>= 8
+        x = ri * we_double[idx]
+        if (ri < ke_double[idx]):
+            return x
+        else:
+            if idx == 0:
+                return ziggurat_exp_r - np.log1p(-next_double(bitgen))
+            elif ((fe_double[idx - 1] - fe_double[idx]) * next_double(bitgen) +
+                  fe_double[idx] < np.exp(-x)):
+                return x
+
+
+@register_jitable
+def random_standard_exponential_f(bitgen):
+    while 1:
+        ri = next_uint32(bitgen)
+        ri >>= 1
+        idx = ri & 0xFF
+        ri >>= 8
+        x = ri * we_float[idx]
+        if (ri < ke_float[idx]):
+            return x
+        else:
+            if (idx == 0):
+                return ziggurat_exp_r_f - np.log1p(-next_float(bitgen))
+            elif ((fe_float[idx - 1] - fe_float[idx]) * next_float(bitgen) +
+                  fe_float[idx] < np.exp(-x)):
+                return x
+
+
+@register_jitable
+def random_standard_exponential_inv(bitgen):
+    return -np.log1p(-next_double(bitgen))
+
+
+@register_jitable
+def random_standard_exponential_inv_fill_f(bitgen):
+    return -np.log1p(-next_float(bitgen))
+
+
+@register_jitable
+def random_standard_gamma(bitgen, shape):
+    if (shape == 1.0):
+        return random_standard_exponential(bitgen)
+    elif (shape == 0.0):
+        return 0.0
+    elif (shape < 1.0):
+        while 1:
+            U = next_double(bitgen)
+            V = random_standard_exponential(bitgen)
+            if (U <= 1.0 - shape):
+                X = pow(U, 1. / shape)
+                if (X <= V):
+                    return X
+            else:
+                Y = -np.log((1 - U) / shape)
+                X = pow(1.0 - shape + shape * Y, 1. / shape)
+                if (X <= (V + Y)):
+                    return X
+    else:
+        b = shape - 1. / 3.
+        c = 1. / np.sqrt(9 * b)
+        while 1:
+            while 1:
+                X = random_standard_normal(bitgen)
+                V = 1.0 + c * X
+                if (V <= 0.0):
```

Comment:
```suggestion
                if (V > 0.0):
```
I think this condition is possibly incorrect. The current implementation reads like "only exit this loop if V is less than or equal to zero". The C implementation:
https://github.com/numpy/numpy/blob/2f1410d3f2bc1058044ec0f1906e24b7f14e56a5/numpy/random/src/distributions/distributions.c#L250-L253
```C
      do {
        X = random_standard_normal(bitgen_state);
        V = 1.0 + c * X;
      } while (V <= 0.0);
```
reads like "only run this loop while V is less than or equal to zero", i.e. exit if V is greater than zero. Essentially the exit condition is inverted? Do you agree? With this change made the `standard_gamma` implementation seems to work.

Reply:
Yeah, that sounds correct; good catch!